src/planner/helpers.ts

```diff
@@ -60,3 +62,66 @@ export function createGlobalConfigurationFromEntities(entities?: EntityList, con
 
     return _.isEmpty(defaultConfig) ? undefined : defaultConfig;
 }
+
+/**
+ * Removes connections that point to non-existing resources and duplicate connections.
+ *
+ * Will not remove connections that point to non-existing block definitions as those are assumed
+ * to exist but not yet loaded.
+ */
+export function cleanupConnections(plan: Plan, blockAssets: AssetInfo<BlockDefinition>[]) {
+    let anyDangling = false;
+    const instanceBlocks: Record<string, AssetInfo<BlockDefinition>> = {};
+    plan.spec.blocks?.forEach((instance) => {
+        const blockRef = parseKapetaUri(instance.block.ref);
+        const asset = blockAssets.find((asset) => parseKapetaUri(asset.ref).equals(blockRef));
+        if (!asset) {
+            return;
+        }
+        instanceBlocks[instance.id] = asset;
+    });
+
+    const connectionSet = new Set<string>();
+
+    const newConnections =
+        plan.spec.connections?.filter((connection) => {
+            const connectionId = `${connection.consumer.blockId}:${connection.consumer.resourceName}:${connection.provider.blockId}:${connection.provider.resourceName}:${connection.port?.type}`;
```

Comment:
So can we have two connections from/to the same resource, but on different ports?

Reply:
Probably not - but dont want to be cursing about why my secondary connection is not created 2 years from now :)